### ISSUE TITLE
Configure group based sorting in agenda

### ISSUE THREAD

**Issue by gerazov** [nvim-orgmode/orgmode#128]
### Does this feature exist in Emacs orgmode core?

N/A

### Feature value

The sorting of tasks in the agenda view seems a bit arbitrary - I get a mix of NEXT and TODO items, and the files they come from are also mixed even if all the tasks are for the same date. I couldn't find how sorting is done in the help file...

For e.g. I get:
```python
понеделник 01 ноември 2021
  work:       Scheduled:  NEXT task name
  nehm:       Scheduled:  NEXT task name
  nehm:       Scheduled:  NEXT task name
  pros:       Scheduled:  NEXT task name
  work:       Scheduled:  NEXT task name
  nehm:       Scheduled:  NEXT task name
  nehm:       Scheduled:  TODO task name
  evoc:       Scheduled:  TODO task name
  nehm:       Scheduled:  NEXT task name
```

It would be very helpful if there is a way to group items based on keywords and files, and possibly set different priorities for sorting them in the agenda, e.g. `DONE > DOING > NEXT > TODO` :+1:

**Comment by kristijanhusak** [comment]
Currently this logic is used: https://orgmode.org/manual/Sorting-of-agenda-items.html

But it seems this should be configurable with this: https://orgmode.org/worg/doc.html#org-agenda-sorting-strategy. It seems that there is no a specific setting to sort by the todo keyword. Both TODO and NEXT keywords are basically a TODO state, and orgmode allows sorting by TODO -> DONE or vice versa.

**Comment by gerazov** [comment]
Ok so it says:

> For the daily/weekly agenda, the items for each day are sorted. The default order is to first collect all items containing an explicit time-of-day specification. These entries are shown at the beginning of the list, as a schedule for the day. After that, items remain grouped in categories, in the sequence given by org-agenda-files. Within each category, items are sorted by priority (see Priorities), which is composed of the base priority (2000 for priority ‘A’, 1000 for ‘B’, and 0 for ‘C’), plus additional increments for overdue scheduled or deadline items.

In orgmode.nvim the specific time todos are listed after the todos without a time of day, this is also less useful. 

And the items are not grouped in categories in my case, even if I pass an explicit list of folders rather than `org/*' they don't get sorted - I get the same mix as before.

:+1: for sorting them by todo state too even if not emacs style :wink:

**Comment by levouh** [comment]
Something like [this](https://d33wubrfki0l68.cloudfront.net/c340243b7ba0a7962d2413038973abda06085b9f/6cd55/ox-hugo/agenda_view.png) (code [here](https://gist.github.com/jethrokuan/78936a44f249e2c1a61b5184669a32d7) and from an article [here](https://blog.jethro.dev/posts/processing_inbox/)) would be something that users would have to add manually, right? Based on the linked code, it appears that `jethrokuan` has items sorted based on their `TODO` state in separate files, then just has a title and list if items that come together to make up a given view.

This looks like it would be something like [this](https://orgmode.org/worg/org-tutorials/org-custom-agenda-commands.html), which would accommodate for not just this use-case but realistically _any_ "custom agenda view" use case.

Not 100% sure of the state right now, but would something like this require changes to the core code (this repo.) @kristijanhusak?

**Comment by kristijanhusak** [comment]
@levouh If we would add some similar configuration, then yeah, it would require some changes, and probably not so small ones. I'm not even sure what's possible with the custom agenda command, I would have to investigate. There is like an AgendaItem that basically wraps all logic around determining what and where to show, so that can be reused, but rendering itself would need changes.

**Comment by levouh** [comment]
Makes sense, I'll see if I can't come up with a good summary before taking a stab at what the code might actually look like. Can open up a separate issue to track it as well, whichever you'd prefer.

The main start on the issue would be:

1. Summary of how the functionality works in _Emacs_
2. What is applicable change-wise here, and what changes should be a part of the core

**Comment by kristijanhusak** [comment]
> Makes sense, I'll see if I can't come up with a good summary before taking a stab at what the code might actually look like. Can open up a separate issue to track it as well, whichever you'd prefer.
> 
> The main start on the issue would be:
> 
> 1. Summary of how the functionality works in _Emacs_
> 2. What is applicable change-wise here, and what changes should be a part of the core

Awesome, thanks! That will help tremendously. You can open up a separate enhancement issue, since this is related only to sorting.

**Comment by gerazov** [comment]
Ok, after some playing around with the sorting in `agenda/init.lua` I've finally fixed this :sweat_smile: 

The problem was that the all the tasks were sorted based on their timestamp. This meant that `date_only` tasks were randomly sorted and all of them went before the tasks scheduled with times of day (I guess `os.time()` was defaulting to 00:00 AM).

Now tasks with time of day are only sorted and they come first in the agenda view, followed by the tasks set without. 

Before (notice how the two tasks from `ea` without time of day get separated):
![image](https://user-images.githubusercontent.com/15214418/143136703-01444165-51d8-4d4d-b4e9-0e38c0c3f63c.png)

After:
![image](https://user-images.githubusercontent.com/15214418/143136367-97bf1fde-50f8-4358-b5b4-697fd4d5ebd2.png)

One thing that I still want to improve is for the tasks to follow the order of the files as given to `orgmode` in the config. This would allow users to e.g. put their personal task `org` last so work tasks comes before or vice versa. To this end I've commented out the file sorting in `parser/files.lua` in `Files.all()`, but still the order is not the same as in the config :thinking:

**Comment by gerazov** [comment]
Ok the ordering is now fixed :love_you_gesture: and the files are still read asyncly.

Here's what I get for the same date (assuming work > ea > valence)
![image](https://user-images.githubusercontent.com/15214418/143144662-82f787f2-b7f7-47b6-8446-5c0511dd23bf.png)

**Comment by kristijanhusak** [comment]
@gerazov can you double check how Emacs orgmode is doing the sorting? I think I adjusted it to work the same way as there, but maybe I missed something.
If it's the same as there, lets go with some solution to allow configuring the sorting as a user, and leave this one as default. As a first (temporary) step we could have a sorting function exposed in the configuration, which you can override to your needs. Once https://github.com/nvim-orgmode/orgmode/issues/135 is done, this function would be removed or changed to adapt the new functionality.

**Comment by gerazov** [comment]
Yeah it's like that in the [specs](https://orgmode.org/manual/Sorting-of-agenda-items.html) as far as I can tell (and it feels right too :sweat_smile: )

> For the daily/weekly agenda, the items for each day are sorted. The default order is to first collect all items containing an explicit time-of-day specification. These entries are shown at the beginning of the list, as a schedule for the day. After that, items remain grouped in categories, in the sequence given by org-agenda-files. Within each category, items are sorted by priority (see Priorities), which is composed of the base priority (2000 for priority ‘A’, 1000 for ‘B’, and 0 for ‘C’), plus additional increments for overdue scheduled or deadline items.

Plus the randomization is minimized to within category listings, which is a bit weird since they are read from the same file :thinking: 

I haven't added the priorities though for the no time-of-day entries, will do that and update. I'm not sure about the 
> additional increments for overdue of scheduled or deadline items.

It reads to me that first comes the overdue deadline, then the overdue scheduled item, then the priority items, then the regular items, no?

I can also implement this for the todo-s list.

**Comment by gerazov** [comment]
Ok, it's now taking deadlines into account, but deadlines are sorted before overdue scheduled items. I'm not sure how it is in emacs as the spec doesn't read very exact. Here's what I get for 3 deadlines (1 overdue, 1 today, 1 tomorrow), 1 overdue scheduled item and 1 today item:

![image](https://user-images.githubusercontent.com/15214418/144678987-800dbf40-9a80-4c09-a561-c8312c09c932.png)

I'm happy with this, how does it look to you guys?

**Comment by kristijanhusak** [comment]
@levouh @gerazov I did some refactoring on agenda to allow easier introduction of custom commands. If you find any issues with current functionality let me know.

@gerazov you will have to rebase your sort PR because I split different agenda view in different files.

**Comment by gerazov** [comment]
Ok will do :+1: 

I'm not sure about the functionality still - I'll have to take a closer look. I get mixing of scheduled and deadlines according to overdue date...

![image](https://user-images.githubusercontent.com/15214418/144858867-d516371c-2536-4b42-a5d1-2bb8abca7353.png)

**Comment by kristijanhusak** [comment]
If you want I'll look into this during this week, you just need to be a bit patient :)
I want to make it 1 to 1 with Emacs orgmode so it's consistent.

**Comment by gerazov** [comment]
Yes please :sweat_smile: no hurry :wink:

**Comment by levouh** [comment]
>I did some refactoring on agenda to allow easier introduction of custom commands. If you find any issues with current functionality let me know

Will do. Once I'm done with the `org-edit-special` stuff I plan on aggregating my notes into #135 and we can go from there. This refactor looks like it will only help, so should be good to go, thanks @kristijanhusak.

**Comment by kristijanhusak** [comment]
@gerazov I pushed a change that should make agenda sorting exactly the same as Emacs in orgmode, at least it was same in my tests.
Please give it a test, and also compare it to the Emacs orgmode to confirm if it's really different.

**Comment by gerazov** [comment]
Ok awesome :sunglasses: I'll give it a go!

**Comment by gerazov** [comment]
I get an empty agenda ... 

![image](https://user-images.githubusercontent.com/15214418/146326994-320a0936-eb80-4049-bd7c-4c96c7d92bbc.png)


Maybe it doesn't find my org files :thinking:  - I list them in my config as such:
```lua
  org_agenda_files = {
      '~/org/work.org',
      '~/org/ea.org',
      '~/org/bme.org',
      '~/org/personal.org',
      },
```

**Comment by kristijanhusak** [comment]
@gerazov did you do `:TSUpdate org` ? Parser got some updates recently.

**Comment by gerazov** [comment]
Tried that now - still same problem. It's fine if I use my fork, so I guess it isn't neovim doing funny business.

**Comment by kris-pappuyar** [comment]
@gerazov try `TSUninstall org` and then `TSInstall org` instead of update. Some people needed to do it that way, update wasn't enough.
I didn't do any bigger changes regarding the agenda fetching, I just updated the sorting which should still return the results.

**Comment by kristijanhusak** [comment]
@gerazov Try doing `TSUninstall org` and then `TSInstall org`, update wasn't sufficient for some people.

**Comment by gerazov** [comment]
Yup works! :pray: 

The sorting is not ok though - the tasks that do not have a time of day stamp should be sorted by category first:

> For the daily/weekly agenda, the items for each day are sorted. The default order is to first collect all items containing an explicit time-of-day specification. These entries are shown at the beginning of the list, as a schedule for the day. After that, items remain grouped in categories, in the sequence given by org-agenda-files. Within each category, items are sorted by priority (see Priorities), which is composed of the base priority (2000 for priority ‘A’, 1000 for ‘B’, and 0 for ‘C’), plus additional increments for overdue scheduled or deadline items.

**Comment by kristijanhusak** [comment]
I would suggest to set up basic emacs installation and compare with that. I did 1 to 1 comparison while setting up the sorting. Maybe what is written and what really happens is not exactly the same.

**Comment by gerazov** [comment]
That doesn't seem right. Did you try with multiple categories? 

Now the overdue items get sorted before today's items, which let's say is ok, but not ideal (and not following specs). Still, even with all the scheduled items for today I get a mix of categories. This is surely not useful (i.e. I can't imagine what are they prioritized with?) And things really gets messy once you have 10+ org files and the tasks from different categories get mixed.

Here's a glimpse (these are all tasks scheduled for today :sweat_smile:):
![image](https://user-images.githubusercontent.com/15214418/146363369-b7e11e72-d2ac-4f45-ae4f-25cd545c19ae.png)

I implemented category based sorting in my [fork](https://github.com/gerazov/orgmode/blob/3ef9f81d2b57135d26ac42c6e2665ede0f13c2e2/lua/orgmode/agenda/init.lua#L63) and it works well:
```lua
    -- if different categories sort by category
    if a.headline:get_category() ~= b.headline:get_category() then
      return category_inds[a.headline:get_category()] < category_inds[b.headline:get_category()]
    end
```

**Comment by kristijanhusak** [comment]
Is this a screenshot from Neovim or emacs Orgmode?
For me sorting by the schedule/deadline date has more sense then to sort by category. Category is just some kind of label to me, doesn't really give any value to the task itself. If I'm overdue on something, I'd like to see that first. 
I didn't think of this sorting functionality myself, as I said, I did 1 to 1 copy from Emacs orgmode. They have more options to offer beside this one, but this is the default (I think called "priority-up"). I think they also support setting up custom sorting function, so I'll probably expose something similar so you can override it as you wish.

**Comment by gerazov** [comment]
Yeah I can go with the overdue priority sorting, but the sorting for tasks scheduled for the same day (with no/equal priorities) should be category based no?

The screenshot is with neovim.

**Comment by kristijanhusak** [comment]
I think it just uses the default sorting, which is as things appear in the files.
It would be best to compare both of them since you have bunch of files, and if you find any differences let me know.

**Comment by gerazov** [comment]
I've given it another go - rebasing and squashing. I can't figure out how to apply the category sorting to the same date items only :sweat_smile:

**Comment by gerazov** [comment]
Ok installing emacs and configuring orgmode seems quite involved - that's how I came here in the first place :sweat_smile: 

What I've done is create a minimum working example.

Here are three identical files having 4 tasks scheduled for today. I had to rename them to `.md` - GitHub doesn't like orgmode apparently :smile:

```md
* TODO Task 1
  SCHEDULED: <2021-12-17 пет>
* TODO Task 2
  SCHEDULED: <2021-12-17 пет>
* TODO Task 3
  SCHEDULED: <2021-12-17 пет>
* TODO Task 4
  SCHEDULED: <2021-12-17 пет>
```

[test1.md](https://github.com/nvim-orgmode/orgmode/files/7736339/test1.md)
[test2.md](https://github.com/nvim-orgmode/orgmode/files/7736335/test2.md)
[test3.md](https://github.com/nvim-orgmode/orgmode/files/7736341/test3.md)

Now when I run the agenda it shows:

![image](https://user-images.githubusercontent.com/15214418/146573880-97e0ef55-2fe6-46e3-8c15-dcf97b05c143.png)

I don't know if emacs reproduces this, but this mixing of categories and tasks within categories doesn't make sense. In fact, it's counter productive.

Also based on this MWE I'm not sure if it's the async loading doing it anymore - it seems random (yet refreshing it doesn't change the ordering) :thinking:

**Comment by gerazov** [comment]
I've made the category based sorting in the PR, but it follows the spec, i.e. sorts by category and then by overdueness. (within category I still get the weird mix but this doesn't bother me as the position of the task in the file is not that important).

![image](https://user-images.githubusercontent.com/15214418/146577637-178a89af-5960-4980-915b-7e5bb90b47a0.png)

I've tried making it work overdue > category but can't make it work ...

The expanded files have three more old tasks:
```md
* TODO Task 1
  SCHEDULED: <2021-12-17 пет>
* TODO Task 2
  SCHEDULED: <2021-12-17 пет>
* TODO Task 3
  SCHEDULED: <2021-12-17 пет>
* TODO Task 4
  SCHEDULED: <2021-12-17 пет>
* TODO Task 5
  SCHEDULED: <2021-12-16 чет>
* TODO Task 6
  SCHEDULED: <2021-12-15 сре>
* TODO Task 7
  SCHEDULED: <2021-12-14 вто>
```

They are here:
[test1.md](https://github.com/nvim-orgmode/orgmode/files/7736550/test1.md)
[test2.md](https://github.com/nvim-orgmode/orgmode/files/7736551/test2.md)
[test3.md](https://github.com/nvim-orgmode/orgmode/files/7736552/test3.md)

**Comment by gerazov** [comment]
Ok with 7ba698a it now works based on overdues + categories:

![image](https://user-images.githubusercontent.com/15214418/146599756-484d0b0e-7712-4906-9a78-5eca388ccca2.png)

**Comment by gerazov** [comment]
I've also added deadlines in the mix. At first it's mixing them with the regular items:

![image](https://user-images.githubusercontent.com/15214418/146600999-10491168-b640-4471-aecc-5fd644301440.png)

But with https://github.com/nvim-orgmode/orgmode/pull/187/commits/f6678a4b6a5efd09cad3d958665f79dd00942466 it's ok :love_you_gesture: 

![image](https://user-images.githubusercontent.com/15214418/146601396-3102d8b1-2ce7-4f45-a8e2-98ef920484f2.png)

**Comment by kristijanhusak** [comment]
@gerazov I tested your files with emacs orgmode, and you were right, it sorts by category in these cases. I pushed a fix where sorting falls back to the default order of headlines being loaded. Pull latest master and give it a test.

**Comment by jgollenz** [comment]
@gerazov is there anything left to this issue or can I close it?

**Comment by gerazov** [comment]
I guess so :sweat_smile: closing ...
